spec/lib/marloss/store_spec.rb

```diff
@@ -18,7 +21,7 @@
   before do
     allow(Aws::DynamoDB::Client).to receive(:new).with(client_options)
       .and_return(ddb_client)
-    allow(store).to receive(:`).with("hostname").and_return(hostname)
+    allow_any_instance_of(described_class).to receive(:`).with("hostname").and_return(hostname)
```

Comment:
why do we need this? we only instantiate the class once at line 14 `let(:store) { described_class.new(table, hash_key, options) }` i don't understand why we need to use `allow_any_instance_of`

Reply:
When calling `host_process_id` from `initialize`, here's what I get without it:
```
  1) Marloss::Store.create_lock should create the lock
     Failure/Error:
       client.put_item(
         table_name: table,
         item: {
           hash_key => name,
           "ProcessID" => process_id,
           expires_key => (Time.now + ttl).to_i
         },
         expression_attribute_names: {
           "#E" => expires_key,
           "#P" => "ProcessID"

       #<InstanceDouble(Aws::DynamoDB::Client) (anonymous)> received :put_item with unexpected arguments
         expected: ({:condition_expression=>"attribute_not_exists(LockID) OR #E < :now OR #P = :process_id", :expression_...>1549560099, "LockID"=>"my_resource", "ProcessID"=>"hostname.local:86776"}, :table_name=>"my_table"})
              got: ({:condition_expression=>"attribute_not_exists(LockID) OR #E < :now OR #P = :process_id", :expression_...=>1549560099, "LockID"=>"my_resource", "ProcessID"=>"doxbook.local:86776"}, :table_name=>"my_table"})
       Diff:

       @@ -2,10 +2,10 @@
           "attribute_not_exists(LockID) OR #E < :now OR #P = :process_id",
          :expression_attribute_names=>{"#E"=>"Expires", "#P"=>"ProcessID"},
          :expression_attribute_values=>
       -   {":now"=>1549560089, ":process_id"=>"hostname.local:86776"},
       +   {":now"=>1549560089, ":process_id"=>"doxbook.local:86776"},
          :item=>
           {"Expires"=>1549560099,
            "LockID"=>"my_resource",
       -    "ProcessID"=>"hostname.local:86776"},
       +    "ProcessID"=>"doxbook.local:86776"},
          :table_name=>"my_table"}]
```

Reply:
Had a more experienced rubyist take a look.. when you do `allow(store)` that calls the `let(:store)`, that instantiates the object but at that point the stub method hasn't been created because it's evaluating the `store` value to pass to `allow()`

Reply:
gotcha. i pulled down the fork and ran it myself and yes is true, because we now call `host_process_id` at initialisation the system call get executed before the stubbing. makes sense now 👍 
i just think `allow_instance_of` tend to make specs flaky and cause side effect - i like to avoid using it when possible but in this case is 🆗 
thanks for trying this out 🙇